### PR TITLE
Migrate template to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Part of #71511

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/deferred_components_prebuild_validator.dart [hunk 1]
```diff
@@ -267,7 +267,7 @@ class _DeferredComponentAndroidFiles {
         templateRenderer: globals.templateRenderer,
       );
     }
-    final Map<String, dynamic> context = <String, dynamic>{
+    final Map<String, Object> context = <String, dynamic>{
```
Comment (campovski) [inline packages/flutter_tools/lib/src/android/deferred_components_prebuild_validator.dart:270]: Change `dynamic` to `Object` in RHS as well?

File: packages/flutter_tools/lib/src/template.dart [hunk 2]
```diff
@@ -120,57 +115,57 @@ class Template {
...
       if (match != null) {
-        final String platform = match.group(1);
-        final String language = context['${platform}Language'] as String;
+        final String? platform = match.group(1);
+        final String? language = context['${platform}Language'] as String?;
         if (language != match.group(2)) {
           return null;
         }
-        relativeDestinationPath = relativeDestinationPath.replaceAll('$platform-$language.tmpl', platform);
+        relativeDestinationPath = relativeDestinationPath.replaceAll('$platform-$language.tmpl', platform!);
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/template.dart:126]: Force unwrap `platform` will crash if `platform` is null.

File: packages/flutter_tools/lib/src/template.dart [hunk 3]
```diff
@@ -120,57 +115,57 @@ class Template {
...
       }
 
-      final bool android = context['android'] as bool;
+      final bool android = (context['android'] as bool?) == true;
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/template.dart:129]: Just curious, if `(context['android'] as bool?)` is null, is `null == true` legal? Does `null == true` just return false?

Based on the previous code, I think it is safe to just force unwrap here. The previous wasn't expect a null value. So maybe:
```
final bool android = context['android']! as bool;
```

File: packages/flutter_tools/lib/src/template.dart [hunk 4]
```diff
@@ -120,57 +115,57 @@ class Template {
...
     ///
     /// Returns null if the given raw destination path has been filtered.
-    String renderPath(String relativeDestinationPath) {
-      final Match match = _kTemplateLanguageVariant.matchAsPrefix(relativeDestinationPath);
+    String? renderPath(String relativeDestinationPath) {
+      final Match? match = _kTemplateLanguageVariant.matchAsPrefix(relativeDestinationPath);
       if (match != null) {
-        final String platform = match.group(1);
-        final String language = context['${platform}Language'] as String;
+        final String? platform = match.group(1);
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/template.dart:121]: If we are sure `platform` is non-null, we should force unwrap here:
```
final String platform = match.group(1)!;
```

Comment (jmagman) [reply packages/flutter_tools/lib/src/template.dart:121]: Good idea.

File: packages/flutter_tools/lib/src/template.dart (hunk 2 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/template.dart:126]: `match.group(1)` can't be null if `match` isn't null, it means there was at least one match.  And it's 1 indexed.

File: packages/flutter_tools/lib/src/template.dart (hunk 3 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/template.dart:129]: There's no guarantee `android` is set.  Like it's not set here:
https://github.com/flutter/flutter/blob/91ed4b2082d1b6b847ade39d94bcaad48bb8f722/packages/flutter_tools/lib/src/project.dart#L787-L793

So if it's not passed in then `context['android']` is null.  So if it's null or false, it should bail from this method and return `null`.  Right?

Comment (cyanglaz) [reply packages/flutter_tools/lib/src/template.dart:129]: Ah i see, you are right. Maybe `context['android'] as bool? ?? false` is a clearer way to do it? Your call tho.

Comment (jmagman) [reply packages/flutter_tools/lib/src/template.dart:129]: We have ` == true` all over this code to check if it's true and not null or false.

**Review by cyanglaz — APPROVED** [review]
LGTM!

**Review by campovski — APPROVED** [review]
LGTM 👍

**Pull request merged**